myia/vm.py

```diff
@@ -0,0 +1,222 @@
+            super().__init__([], frame, None)
+
+    def __init__(self, graph: Graph, args: List[Any], parent: 'VMFrame',
+                 closure: 'VMFrame' = None) -> None:
+        """Build a frame."""
+        self.graph = graph
+        self.args = args
+        self.parent = parent
+        self.closure = closure
+        self.values: Dict[ANFNode, Any] = dict()
+        self.todo: List[ANFNode] = [self.graph.return_]
+
+    def run(self) -> 'VMFrame':
+        """Run the graph to completion and return the result."""
+        while True:
+            try:
+                res = self.advance()
+            except StopIteration:
+                # Return is handled here because the flow control is a bit
+                # too complex for the parent frame to get a value back
+                # from the called frame.
+                if self.parent is not None:
+                    self.parent.values[self.parent.todo[-1]] = self.result()
+                    self.parent.todo.pop()
+                return self.parent
+            if res is not None:
+                return res
+
+    def tail(self) -> bool:
+        """Are we in a tail call?."""
+        return self.todo[-2] is self.graph.return_
+
+    def done(self) -> bool:
+        """Are we done?."""
+        return len(self.todo) == 0
+
+    def result(self):
+        """Get this frame's result."""
+        return self.values[self.graph.return_]
+
+    def get_value(self, node: ANFNode):
+        """Get the value for a node.
+
+        This will check if we've already computed a value for it and
+        add it to the queue otherwise. If there was no value, we stop
+        the evaluation of the current node and restart the step, since
+        no useful work was done.
+
+        If the value we want is in another graph, we will resume
+        execution in a different frame and jump back to this one later.
+        """
+        frame = self
+        while frame and node.graph and node.graph is not frame.graph:
+            frame = frame.closure
+        if node not in frame.values:
+            # we set up a fake node to jump back to this frame after
+            # the value that we need has been evaluated.
+            if frame is not self:
+                frame.todo.append(self.Jump(self))
+            frame.todo.append(node)
+            if frame is not self:
+                raise self.Return(frame)
+            else:
+                raise self.Continue()
+        return frame.values[node]
+
+    def wrap_closure(self, value):
+        """Wrap graphs that are closures."""
+        if isinstance(value, Graph):
+            if any(n.graph and n.graph is not value
+                   for n in dfs(value.return_)):
+                return self.Closure(value, self)
+        return value
+
+    def do_call(self, graph: Union[Graph, 'VMFrame.Closure'], args: List[Any]):
+        """Perform a call.
+
+        This will handle tail calls and closures.
+
+        """
```

Comment:
You should check that there are exactly as many arguments as the graph takes parameters. This would have caught the bug in https://github.com/mila-udem/myia/pull/4#issuecomment-358711256 immediately, and it is virtually certain that it will catch other similar bugs in the future.

Reply:
added